Added check for menu router, when adding in a default itemid.

tonypartridge opened the pull request:
Pull Request for Issue #15730.

### Summary of Changes

Add check that current is not the homepage menu itemid, then check the user has access before applying the Itemid. 

### Testing Instructions

Before patch, set the homepage as a 'Registered' Access level menu on a blank install. And access the menu, any menu item will do.


### Expected result

You are redirected to a login page.

### Actual result

You end up in a redirect loop.

### Documentation Changes Required

Commit:
Added check for menu router, when adding in a default itemid.

zero-24 commented:
@tonypartridge can you take a look into drone? http://213.160.72.75/joomla/joomla-cms/4442 ?

Commit:
Fix spacing for drone.

tonypartridge commented:
@zero-24 I was doing it just as you commented :-). All done, waiting for it to be re-tested it.

Commit:
Fix spacing for drone.

Commit:
Fix spacing for drone.

mbabker commented on libraries/cms/component/router/rules/menu.php:
```diff
@@ -130,7 +130,12 @@ public function preprocess(&$query)
 
-		if (!empty($default->id))
+		// Get the current users access levels
+		$levels = JAccess::getAuthorisedViewLevels(JFactory::getUser()->id);
```
`JFactory::getUser()->getAuthorisedViewLevels()`

tonypartridge replied:
Updated as per request :-)

Commit:
changed methoed of getting authorised access levels as per @mbabker comment.

wilsonge commented:
Can we please ensure there is a test on a multilanguage site - we might need to add this check to the `if` statement above it as well - just on code review. It might be ok.

Commit:
Fix for handling multi and monolangual websites, thanks to @willsonge for the code/pointer.

wilsonge commented:
I have tested this item :white_check_mark: successfully on 9398f56181af6d099b06327d8302a9fd1ff030ce<br /><br />Tested on a mongolanguage and multilanguage site. Seems to redirect correctly! Nice job!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16197">issues.joomla.org/tracker/joomla-cms/16197</a>.</sub>

wilsonge commented:
@rdeutz we've had multiple trackers about this. I know it's last minute but it really should ship in 3.7.2 (even if that means we delay the release another day so the CMS release team have time)

Quy commented on libraries/cms/component/router/rules/menu.php:
```diff
@@ -132,8 +132,8 @@ public function preprocess(&$query)
-		// Make sure there is an menu item ID, check it is the home menu, and the user can actually access this menu item,
-		// otherwise it can end up in a redirect loop.
+
+		// Make sure there is an menu item ID, check it is the home menu, and the user can actually access this menu item, otherwise it can end up in a redirect loop.
```
Change `an menu` to `a menu`.

Quy commented on libraries/cms/component/router/rules/menu.php:
```diff
@@ -130,7 +130,11 @@ public function preprocess(&$query)
 		$default = $this->router->menu->getDefault($language);
 
-		if (!empty($default->id))
+		//Get the current users access levels
```
Change `users` to `user's`.

Quy commented on libraries/cms/component/router/rules/menu.php:
```diff
@@ -130,10 +130,10 @@ public function preprocess(&$query)
 		$default = $this->router->menu->getDefault($language);
 
-		//Get the current users access levels
+		// Get the current users access levels
```
Change `users` to `user's`.

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 9398f56181af6d099b06327d8302a9fd1ff030ce<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16197">issues.joomla.org/tracker/joomla-cms/16197</a>.</sub>

franz-wohlkoenig commented:
RTC after two successful tests.

rdeutz commented:
It is not enough to make sure this patch fixes a problem we also need to make sure it doesn't has a side effect. This need intensive testing.

Re 3.7.2, we are at the end of the release process, more or less hitting the publish button is missing, so this here is too late for 3.7.2.

infograf768 commented:
Works fine on multilang.

brianteeman committed:
comment typos

fix the typos spotted by @quy - they were not introduced in this PR but still good to fix

brianteeman commented:
I took a 3.6.5 web site and made the home page registered only AND made sure there was no login menu item
Loading the site on the front end I am taken to index.php/component/users/?view=login and can login

I upgraded the site to 3.7.1 and I get the "too many redirects" issue

I applied the patch and once again it is working and I am taken to  index.php/component/users/?view=login and can login

So this is a successful test. 

**NOTE if you have a login menu item then you would not have needed this patch at all**

brianteeman commented:
I have tested this item :white_check_mark: successfully on 18bfbaff54666e08cad08df29348f1cfee0e2cba<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16197">issues.joomla.org/tracker/joomla-cms/16197</a>.</sub>

tonypartridge commented:
@infograf768 could you please add a confirmed test if you have tested it on multilingual sites?

infograf768 commented:
I have tested this item :white_check_mark: successfully on 18bfbaff54666e08cad08df29348f1cfee0e2cba<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16197">issues.joomla.org/tracker/joomla-cms/16197</a>.</sub>

wilsonge commented:
4 good tests - I'm merging this

brianteeman commented:
I really dont see the need to rush this into 3.7.2 when it is easily fixed by anyone right now who has this issue by adding a login menu item and this can be added to the docs

infograf768 commented:
3.7.2 packs are ready. Last night merges as well as this one will be in 3.7.3

tonypartridge commented:
whilst that is true @brianteeman It's about the people who upgrade their sites and have no clue about why their site has suddenly gone down with redirect loops. And another user in the original posts has tons of websites to update, doing so would cause them all to go down when he updates via the update manager.

brianteeman commented:
@wilsonge was trying to get this into 372

wilsonge commented:
We're not merging this into 3.7.2 - hence why I've merged it with a 3.7.3 milestone

brianteeman commented:
@wilsonge our messages crossed in the ether ;)

rdeutz commented:
Wondering why travis didn't run on this, on jenkins one unit test is falling

stevlam commented:
thanks! I see fix will be on 3.7.3 . OK will manually replace and fix the site until them

infograf768 commented:
@tonypartridge 
Please redo your PR. We have to make sure tests are not failing.

tonypartridge commented:
I blame @brianteeman 's language changes ;-) Will do it now.

brianteeman commented:
they were in comments !!
